src/Core/View.js

```diff
@@ -291,8 +295,31 @@ View.prototype.addLayer = function addLayer(layer, parentLayer) {
 View.prototype.notifyChange = function notifyChange(needsRedraw, changeSource) {
     this._changeSources.add(changeSource);
     this.mainLoop.scheduleViewUpdate(this, needsRedraw);
+    if (this.avoidCollision && (changeSource === undefined || changeSource.type != 'Mesh')) {
+        this.checkCollisionWithGeometryLayer();
+    }
+};
+
+
+/**
+ * Test for collision between camera and geometry layer (DTM/DSM) to adjust camera position
+ */
+View.prototype.checkCollisionWithGeometryLayer = function checkCollisionWithGeometryLayer() {
+    const camLocation = new Coordinates(this.referenceCrs, this.camera.camera3D.position).as('EPSG:4326');
+    if (this._layers[0] && this._layers[0] instanceof GeometryLayer) {
+        const elevationUnderCamera = DEMUtils.getElevationValueAt(this._layers[0], camLocation);
+        if (elevationUnderCamera != undefined) {
+            const difElevation = camLocation.altitude() - (elevationUnderCamera.z + this.minDistanceCollision);
+            // We move the camera to avoid collisions if too close to terrain
+            if (difElevation < 0) {
+                camLocation.setAltitude(elevationUnderCamera.z + this.minDistanceCollision);
```

Comment:
I'm not sure all controls instances support external changes of camera's position.